### ISSUE TITLE
DML and IUserType

### ISSUE THREAD

**Issue by Daythril** [nhibernate/nhibernate-core#2650]
I have a database column with string values like 'void' and 'paid'. I use an IUserType to convert those into Enum values. This works just fine, except: I just did a DML Update, as follows

```
session.Query<Invoice>()
    .Where(i => ids.Contains(i.id))
    .Update(i => new Invoice(){status = EnumInvoiceStatus.Void});
```
I had expected the status column in the database to be updated to 'Void', but instead it got set to '4', which is the integer value of `Void` in the Enum.

I can understand that a lot of niceties were foregone to get the efficiency of DML queries, but it should have been easy enough -- I would think -- to have used the IUserType to generate 'Void' from EnumInvoiceStatus.Void when generating the SQL from this query.  As it is I see no way to do this, since I can't assign a string to `status `-- the compiler would reject it.

I haven't seen mention of a similar issue. Is this being worked on? Is there a workaround?

**Comment by maca88** [comment]
> I use an IUserType to convert those into Enum values

There is no need to use an `IUserType` for string enums, you can extend `EnumStringType<EnumInvoiceStatus>` to achieve it. Here is an example:
Class: https://github.com/nhibernate/nhibernate-core/blob/d6e054f2bbfda43fccf82d54b7cda4e5aecbab5d/src/NHibernate.Test/NHSpecificTest/NH1941/SexEnumStringType.cs#L6 overriding `GetValue` and `GetInstance` methods is not required in case the database values are the same as enum names.

Mapping: https://github.com/nhibernate/nhibernate-core/blob/d6e054f2bbfda43fccf82d54b7cda4e5aecbab5d/src/NHibernate.Test/NHSpecificTest/NH1941/Mappings.hbm.xml#L11

Can you please share the `IUserType` that you are using, so that we can easier reproduce the issue?

**Comment by Daythril** [comment]
Thanks for your response, and I’m sorry it’s taken me so long to get back to you – I got called away onto other projects.

I was perhaps not clear enough about my problem.  It wasn’t with the conversion of Enums to strings and back, but with the fact that the conversion wasn’t being called as part of the process of going from LINQ to SQL because I was hooking into that process after that had already been done.  If have since solved the problem, and in case you’re interested, I’ll show you.

Previously I had used a class ComputedPropertyGeneratorRegistry : DefaultLinqToHqlGeneratorsRegistryto register mappings between computed properties in my persistent classes and Expressions, which were then processed by the BuildHql override in a class CalculatedPropertyGenerator<T, TResult> : BaseHqlGeneratorForProperty.

But like I said, that was too late in the process of converting LINQ to SQL.  Now I install a custom query-provider (cfg.LinqQueryProvider<CustomQueryProvider>()), which catches the process early enough to do what I need.
